File: recursion/can_jump.js
// return true or false if able to reach the last index
// https://leetcode.com/problems/jump-game/solution/

// Iterative

// function canJump(nums) {
//   var finalIndex = nums.length - 1;
//   var needToReach = nums.length - 1;
//
//   // position set to start comparing the second to last element
//   for (var position = finalIndex- 1; position >= 0; position--) {
//     var value = nums[position];
//     if ((value + position) >= needToReach) {
//       needToReach = position;
//     }
//   }
//
//   return needToReach === 0 ? true : false;
// }
//
// console.log(jump([1,2,3,2,3,0,0,1]) == true);
// console.log(jump([1,2,3,2,3,0,0,0,0]) == false);

// Recursive traversal

function canJump(nums) {
  var needToReach = nums.length - 1;
  var position = needToReach - 1;
  // starting from the second to last
  var jumpTest = canJumpHelper(nums, needToReach, position)
  return jumpTest;
}

function canJumpHelper(nums, needToReach, position) {
  // logic updating passed in data
  var value = nums[position];
  if ((value + position) >= needToReach) {
    needToReach = position;
  }

  // break condition after iterating through all array elements in reverse
  if (position === 0) {
    return needToReach === 0 ? true : false;
  }

  // n - 1 recursive call
  return canJumpHelper(nums, needToReach, position - 1)
}

console.log(canJump([1,0]) === true);
console.log(canJump([0,0]) === false);
console.log(canJump([1,2]) === true);
console.log(canJump([1,2,3]) == true);
// console.log(canJump([1,2,3]) === true);
// console.log(canJump([1,2,3,2,3,0,0,1]) == true);
// console.log(jump([1,2,3,2,3,0,0,0,0]) == false);

File: dynamic_programming/can_jump/can_jump.rb
# https://leetcode.com/problems/jump-game/

# Approach 1: Top down recursive solution

def can_jump(array)
  can_jump_helper(array, array.size-1)
end

def can_jump_helper(array, last_index)
  # base case - index has successfully backtracked to the beginning position
  return true if last_index == 0

  (0..last_index - 1).each do |i|
    if (array[i] + i >= last_index)
      if can_jump_helper(array, last_index - 1)
        return true
      end
    end
  end
  return false
end

# Approach 2: Top down dynamic programming solution

def can_jump(array)
  can_jump_helper(array, array.size-1, {})
end

def can_jump_helper(array, last_index, cache)
  # base case - index has successfully backtracked to the beginning position
  return true if last_index == 0

  (0..last_index - 1).each do |i|
    if (array[i] + i >= last_index)
      if cache[i] || can_jump_helper(array, i, cache)
        cache[i] = true
        return true
      end
    end
    cache[i] = false
  end
  return false
end

p can_jump([2,0,1,3,0]) == true
p can_jump([2,0,1,1,0]) == true
p can_jump([2,0,1,0,0]) == false
